Jetpack Social: Add footer and subscribe flow for prepublishing social accounts

dvdchr opened the pull request:
Refs #20783 

In this PR:

- Added a table footer view, containing the remaining shares label and the subscribe button. This footer is only shown when the site has sharing limits.
- Tapping the subscribe button shows the checkout page in an authenticated web view.
- When dismissed, the view controller will try to sync the sharing limit in case the user did make a purchase. When the user obtains unlimited sharing, the social accounts sheet should be refreshed.

Some previews:

Normal state | Warning state
-|-
![footer_normal](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/ccc00c7e-93dd-430a-952c-185339733c5b) | ![footer_warning](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/1dc1b3a4-7e45-44e8-998a-7fd1f07b3793)

> **Warning**
> Changes in this view are still isolated. In the next PR, the changes will be propagated back to the pre-publishing sheet, and persisted into Core Data.
> 
> Additionally, I've also noticed some oddities with the table view displayed in a popover. Somehow the default top padding of the table is now gone, and I'm not sure what caused it. I'm beginning to think it's a framework thing, but I'll leave it be for now since it's a minor visual issue.

## To test

### Case 1: Testing limited sharing state

- Launch the Jetpack app.
- Switch to a site with limited social sharing.
- Open a draft, or create a new blog post along with some title and text.
- Tap Publish.
- Tap the auto-sharing cell.
- 🔎  Verify that the footer is displayed.

### Case 2: Testing unlimited sharing state

- Repeat the steps in Case 1 with a WP.com site.
- 🔎 Verify that the footer view is NOT displayed.

### Case 3: Testing the warning state

To make things easy, let's make 2 small changes in the `PrepublishingSocialAccountsViewController`:

<details>
<summary>Test setup</summary>

---

In line 68, set all accounts to be disabled:

```swift
.init(service: service.name, account: $0.account, keyringID: $0.keyringID, isOn: $0.enabled)
```

And in line 72, replace the assigned `SharingLimit` with a test instance:

```swift
self.sharingLimit = .init(remaining: 1, limit: 30)
```

---

</details>

- Launch the Jetpack app again, and follow the testing steps in Case 1.
- 🔎  Verify that the remaining shares text is shown with a warning style.

### Case 4: Testing the subscribe flow

In addition to the test setup above, let's now going to make additional changes to simplify testing:

<details>

<summary>Additional test setup</summary>

---

Replace the entire body of `onCheckoutDismissed()` in line 261-276 with:

```swift
        DispatchQueue.main.async { [weak self] in
            self?.sharingLimit = nil // assume the upgrade is successful.
            self?.toggleInteractivityIfNeeded()
            self?.tableView.reloadData()
        }
```

This simply assumes that the upgrade is successful (sharing limit is removed), and the table view should be reloaded properly.

---

</details>

- With the test setups set, launch the Jetpack app and follow the testing steps from Case 1.
- Enable one of the accounts. The other account should now be in a disabled/non-interactive state.
- Tap the Subscribe button.
- Dismiss the web view.
- 🔎 Verify that the footer view is now gone.
- 🔎 Verify that the account switches can be interacted with.

### Additional

- Test with accessibility font size.
- Test with landscape orientation.
- Test with iPad.

## Regression Notes
1. Potential unintended areas of impact
Should be none. The feature is unreleased.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

3. What automated tests I added (or what prevented me from doing so)
N/A.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] VoiceOver.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dvdchr committed:
Add table footer view

dvdchr committed:
Show footer view in social accounts screen

dvdchr committed:
Manually configure content size calculation

dvdchr committed:
Show checkout page when subscribe button is tapped

dvdchr committed:
Switch from post object ID to Blog ID

dvdchr committed:
Sync sharing limit when checkout webview is dismissed

dvdchr committed:
Improve accessibility for the remaining text component

dvdchr committed:
Use SwiftUI.Font instead of converting from UIFont

Converting from UIFont doesn't automatically scales based on
content size category, and there's no `adjustsFontForContentSizeCategory`
yet in SwiftUI.

dvdchr committed:
Add a constant bottom inset for iPad to prevent weird resizing

wargcm commented on WordPress/Classes/ViewRelated/Post/Prepublishing/PrepublishingSocialAccountsViewController.swift:
```diff
@@ -172,6 +228,52 @@ private extension PrepublishingSocialAccountsViewController {
+                break
+            }
+        }
+    }
```
I believe the sharing limit still returns for blogs which have upgraded (pctCYC-PK-p2#getting-the-sharing-limit-information).

Wouldn't we instead need to fetch the blog's features and then check the `sharingLimit` property on the `Blog` object? That way it could check if the blog has `social-shares-1000`.

dvdchr replied:
Ah, you're right. Good catch! I've updated this in https://github.com/wordpress-mobile/WordPress-iOS/pull/21179/commits/fbd98d9e1e68f6f07b1fd39736f12ece55133e4b to sync the blog instead of the sharing limit information.

Ready for your second round of review, please!

dvdchr committed:
Sync blog instead of just the sharing limit

The view controller now stores `BlogService` instead of `JetpackSocialService`
to perform `syncBlog` when checkout is dismissed.

Additionally, the property `isSharingLimitReached` is updated to
`canInteractWithDisabledConnections`, and its value is now initialized
in the `init` method according to the sharing limit. This takes care
of an edge case where the remaining shares is 0, where all accounts
should've been disabled.

Lastly, added some docs and minor code organization.

wargcm commented on WordPress/Classes/ViewRelated/Post/Prepublishing/PrepublishingSocialAccountsViewController.swift:
```diff
@@ -48,14 +71,21 @@ class PrepublishingSocialAccountsViewController: UITableViewController {
         }
         self.shareMessage = model.message
-        self.sharingLimit = model.sharingLimit
+        self.sharingLimit = .init(remaining: 1, limit: 30)
```
Looks like a testing change made it in here.
```suggestion
        self.sharingLimit = model.sharingLimit
```

dvdchr replied:
🤦🏼‍♂️ of course. Thanks!

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21179/commits/d72d82fcb590fdbb6e180499110274487adc9212.

wargcm commented on WordPress/Classes/ViewRelated/Post/Prepublishing/PrepublishingSocialAccountsTableFooterView.swift:
```diff
@@ -0,0 +1,108 @@
+
+        static let remainingSharesLabelTextFormat = NSLocalizedString(
+            "prepublishing.socialAccounts.footer.remainingShares.text",
+            value: "%1$d social shares remaining in the next 30 days",
```
Not an issue for this PR but I wonder if we should create a plural and non-plural version of this string now that we've removed `/<ShareLimit>` from it.

dvdchr replied:
Good point. I'll consider this in the next PR!

wargcm reviewed (approved):
✅ All test cases pass. Additionally, I checked the new blog sync changes with Charles and it worked as expected 🎉 .

There was only a minor issue with some test data and a note on a string for us to consider.

LGTM! :shipit:

dvdchr committed:
Remove test code
